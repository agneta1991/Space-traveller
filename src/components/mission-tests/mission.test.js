import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { toBeInTheDocument, toHaveTextContent } from '@testing-library/jest-dom/matchers'; // Import matchers
import Missions from '../missions';

expect.extend({ toBeInTheDocument, toHaveTextContent });

const mockStore = configureStore([]);

describe('Missions Component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      missions: {
        data: [
          {
            mission_id: '1',
            mission_name: 'Test Mission 1',
            description: 'Test Description 1',
            reserved: false,
          },
          {
            mission_id: '2',
            mission_name: 'Test Mission 2',
            description: 'Test Description 2',
            reserved: true,
          },
        ],
        status: 'succeeded',
        error: null,
      },
    });
  });

  it('renders loading state', () => {
    store = mockStore({
      missions: {
        data: [],
        status: 'loading',
        error: null,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders error state', () => {
    store = mockStore({
      missions: {
        data: [],
        status: 'failed',
        error: 'Test Error',
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders mission data and handles joining/leaving', async () => {
    const { container, getByText } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    const joinButton = getByText('Join Mission');
    fireEvent.click(joinButton);

    await waitFor(() => {
      const updatedJoinButton = getByText('Leave Mission');
      expect(updatedJoinButton).toBeInTheDocument();
    });

    const leaveButton = getByText('Leave Mission');
    fireEvent.click(leaveButton);

    await waitFor(() => {
      const updatedLeaveButton = getByText('Join Mission');
      expect(updatedLeaveButton).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });
});
