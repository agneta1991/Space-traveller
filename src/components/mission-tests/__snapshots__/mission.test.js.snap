// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Component renders error state 1`] = `
<div>
  <div>
    Error loading missions data.
  </div>
</div>
`;

exports[`Missions Component renders loading state 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`Missions Component renders mission data and handles joining/leaving 1`] = `
<div>
  <table
    class="missions-table"
  >
    <thead>
      <tr>
        <th
          class="missionColumn"
        >
          Mission
        </th>
        <th
          class="wider-column"
        >
          Description
        </th>
        <th>
          Status
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr
        class="even-row"
      >
        <td
          class="bold-text"
        >
          Test Mission 1
        </td>
        <td>
          Test Description 1
        </td>
        <td>
          <p
            class="status"
            data-testid="mission-status"
            style="background-color: rgb(109, 117, 125);"
          >
            NOT A MEMBER
          </p>
        </td>
        <td
          class="joinContainer"
        >
          <button
            class="joinMission"
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
      <tr
        class="odd-row"
      >
        <td
          class="bold-text"
        >
          Test Mission 2
        </td>
        <td>
          Test Description 2
        </td>
        <td>
          <p
            class="status"
            data-testid="mission-status"
            style="background-color: rgb(65, 155, 249);"
          >
            ACTIVE MEMBER
          </p>
        </td>
        <td
          class="joinContainer"
        >
          <button
            class="joinMission"
            style="color: red; border: 1px solid red;"
            type="button"
          >
            Leave Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
